Parser/pegen/pegen.c

```diff
@@ -530,13 +540,21 @@ _PyPegen_fill_token(Parser *p)
 
     if (p->fill == p->size) {
         int newsize = p->size * 2;
-        p->tokens = PyMem_Realloc(p->tokens, newsize * sizeof(Token *));
-        if (p->tokens == NULL) {
-            PyErr_Format(PyExc_MemoryError, "Realloc tokens failed");
+        Token **new_tokens = PyMem_Realloc(p->tokens, newsize * sizeof(Token *));
+        if (new_tokens == NULL) {
+            PyMem_Free(p->tokens);
+            p->tokens = NULL;
             return -1;
         }
+        else {
+            p->tokens = new_tokens;
+        }
         for (int i = p->size; i < newsize; i++) {
             p->tokens[i] = PyMem_Malloc(sizeof(Token));
+            if (p->tokens[i] == NULL) {
```

Comment:
You need to update  p->size here. Maybe set it it i or i-1, I'm not sure :-p

Reply:
Is that needed if we fail hard here? If `fill_token` returns -1, the parser fails for every rule until it backtracks out the start function and it destroys the `Parser` struct.

Reply:
When the parser is destroyed, you have to free memory of tokens between `p->size` and `<i value when the loop stopped>`, not only `0` to `p->size`.

Reply:
Of course!